corva/middleware/unpack_context.py

```diff
@@ -0,0 +1,49 @@
+import inspect
+from typing import Callable, Optional, Tuple
+
+from pydantic.utils import lenient_issubclass
+
+from corva.models.base import BaseContext, BaseEvent
+from corva.network.api import Api
+from corva.state.redis_state import RedisState
+
+
+def _parse_call(call: Callable) -> Tuple[Optional[str], Optional[str], Optional[str], Optional[str]]:
```

Comment:
what is the purpose of this function?

Reply:
it scans parameter annotations of user function and finds names of parameters which have needed annotations. allows this functionality [link](https://github.com/corva-ai/python-sdk/blob/7e8a5aa1ceaf37e7447693b05cbc8f6c4ae0459a/docs_src/tutorial_1_hello_world.py#L7)

Reply:
inspired by [this](https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L336)

Reply:
we don't need to parse it. We already know what is passed into this function

Reply:
I'll give some examples, why unpack_context mware is needed

1. different app types need different parameter sets. e.g task apps don't need `State` param, because they are only called once

```python3
@app.task
def user_job(event: Event, api: Api): pass
```

2. parameters may be shuffled e.g

```python3
@app.stream
def user_job(api: Api, state: State, event: Event): pass
``` 

3. I also wanted to add support to pass additional `Context` param. which may not be always needed

```python3
@app.stream
def user_job(api: Api, state: State, event: Event, context: Context): pass
```

Reply:
Different types have different logic, and that is ok
If these are positional arguments - they should not be shuffled
If these parameters are named arguments - we don't care if they are shuffled
The context could be added as an optional named parameter

From my perspective, the whole `unpack` module resolves the non-existing problem

Reply:
If there is no `unpack` middleware, what do we do in the next cases?

> If these are positional arguments - they should not be shuffled

1.  arguments not shuffled but in second case state is missing

```python3
@app.stream
def user_job(event: Event, state: State, api: Api): pass

@app.stream
def user_job(event: Event, api: Api): pass
```

> If these parameters are named arguments - we don't care if they are shuffled

2. user uses different parameter name. it is probably not good to make user memorize parameter names.

```python3
@app.stream
def user_job(e: Event, s: State, a: Api): pass
```

Reply:
> arguments not shuffled, but in the second case, the state is missing

They should not be missing. Why are they missing?

> user uses different parameter names. it is probably not good to make user memorize parameter names

We are sending 3 parameters into the function: `event,` `state,` and `api,` in the same order. For example, lambda does this with `event` and `context. It's lambda's responsibility to send them. But it's my responsibility, as a developer, to receive them. I can call my function whatever I want, I can call my arguments whatever I want, but the first argument will be an event, and the second one - the context. No matter how I will call them in my function. Can we do that?.`

Reply:
> They should not be missing. Why are they missing?

different app types need different parameter sets. e.g task apps don't need State param, because they are only called once

Reply:
But we have different decorators for different types. At the execution time, we know what parameters which function should receive

Reply:
Reworked unpack_context.py to always pass predefined parameters like `event` and `context` in aws lambda functions.